### PR TITLE
party map layout fixes and tweaks

### PR DESCRIPTION
**Pull request by denisdimitrov**
- increase navbar zindex to show it above schedule
- add new chatbox layout to private message modal
- add padding to number of partygoers online
- render room count conditionally to avoid rendering blue box
- open external internal links differently

This PR is pending on #845 and is going to have breaking changes + conflicts

### CODE REVIEW

File: src/components/molecules/NavBar/playa.scss [hunk 1]
```diff
@@ -16,7 +16,7 @@ $sand: #937c63;
 
 .navbar.navbar_playa {
   position: fixed;
-  z-index: 100;
+  z-index: 301;
```
Comment (thecadams) [inline src/components/molecules/NavBar/playa.scss:19]: Any reason for `301` specifically... let's move this zindex to a variable, and aim to move more later so we have them all in one place.

File: src/components/templates/PartyMap/components/Map/Map.scss [hunk 2]
```diff
@@ -279,6 +279,8 @@ $border-radius: 28px;
     transition: all 400ms cubic-bezier(0.23, 1, 0.32, 1);
     opacity: 0;
     transform: translateY(-20px) scale(0.8);
+    min-width: 100px;
+    max-width: 120px;
```
Comment (thecadams) [inline src/components/templates/PartyMap/components/Map/Map.scss:283]: just wondering - this seems suspect; was it tested at many widths?

File: src/components/templates/PartyMap/components/Map/Map.scss [hunk 3]
```diff
@@ -332,6 +334,8 @@ $border-radius: 28px;
     transition: height 400ms cubic-bezier(0.23, 1, 0.32, 1);
     display: block;
     transition: transform 600ms cubic-bezier(0.23, 1, 0.32, 1);
+    min-width: 150px;
+    max-width: 280px;
```
Comment (thecadams) [inline src/components/templates/PartyMap/components/Map/Map.scss:338]: same here - wondering if this has been tested at many widths

File: src/utils/url.ts [hunk 4]
```diff
@@ -39,7 +39,11 @@ export const openRoomUrl = (url: string) => {
 };
 
 export const openUrl = (url: string) => {
-  window.open(url, "_blank", "noopener,noreferrer");
+  if (!isExternalUrl(url)) {
+    window.open(url, "_blank", "noopener,noreferrer");
```
Comment (thecadams) [inline src/utils/url.ts:43]: nit: let's use the utility function for this

**Review by thecadams — COMMENTED** [review]
Looks good - only a few minor questions.

File: src/components/molecules/VenuePartygoers/VenuePartygoers.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+.attendees-container {
```
Comment (0xdevalias) [inline src/components/molecules/VenuePartygoers/VenuePartygoers.scss:1]: Should this be scoped under something `VenuePartygoers` related?

Generally in the past I've used patterns where the top 'container' uses a 'wrapper' class (eg. `.venue-partygoers`), and then all other more specific rules live under this. That way they will only be applied when they appear within the 'wrapping classes scope'.

That way we can be more confident that our CSS styling doesn't weirdly change/break other things, without having to result to a CSS-in-JS pattern (which in many cases feels heavy/overkill IMO)

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 6]
```diff
@@ -61,10 +67,29 @@ const UserProfileModal: React.FunctionComponent<PropTypes> = ({
...
 
+  const chatContext = useContext(ChatContext);
+
+  const submitMessage = useCallback(
+    async (data: { messageToTheBand: string }) => {
+      chatContext &&
+        user &&
+        chatContext.sendPrivateChat(
+          user.uid,
+          fullUserProfile!.id,
```
Comment (0xdevalias) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:78]: Do we want to be using `!` here?

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 7]
```diff
@@ -61,10 +67,29 @@ const UserProfileModal: React.FunctionComponent<PropTypes> = ({
     return undefined;
   }, [rest]);
 
+  const chatContext = useContext(ChatContext);
+
+  const submitMessage = useCallback(
+    async (data: { messageToTheBand: string }) => {
+      chatContext &&
+        user &&
```
Comment (0xdevalias) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:75]:
```suggestion
      if (!chatContext || !user) return
      
```

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 8]
```diff
@@ -61,10 +67,29 @@ const UserProfileModal: React.FunctionComponent<PropTypes> = ({
...
+    [chatContext, fullUserProfile, user]
+  );
+
   if (!fullUserProfile || !fullUserProfile.id || !user) {
     return <></>;
   }
 
+  const chats: WithId<PrivateChatMessage>[] = privateChats.filter(
+    (chat) => chat.from === fullUserProfile.id || chat.to === fullUserProfile.id
+  );
```
Comment (0xdevalias) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:91]: This may be worth `useMemo`ing

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 9]
```diff
@@ -61,10 +67,29 @@ const UserProfileModal: React.FunctionComponent<PropTypes> = ({
     return undefined;
   }, [rest]);
 
+  const chatContext = useContext(ChatContext);
+
+  const submitMessage = useCallback(
+    async (data: { messageToTheBand: string }) => {
```
Comment (0xdevalias) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:73]: I realise this is what `ChatBox` currently expects, but it feels somewhat needlessly complex/an implementation detail. I wonder if we can refactor this to just take a `msg: string` instead?

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 6 above)
Comment (0xdevalias) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx:78]: Also, what does `fullUserProfile` refer to here?

`chatContext.sendPrivateChat` seems to take `(from, to, text)` as params, implying we are sending a message to `fullUserProfile.id` here?

I'm guessing that since this is passed down from props, the implication is that `fullUserProfile` is the user that this `UserProfileModal` is for/showing. And `user` is the current person that opened the modal.

The ambiguous use of `user` in this component makes it hard to follow. I think renaming `user` to something implying "the current person" (`loggedInUser`, `authenticatedUser`, `myUser`?), and/or renaming `fullUserProfile` to something implying "the person this modal is about" would make this a lot easier to follow.

File: src/utils/url.ts (hunk 4 above)
Comment (0xdevalias) [reply src/utils/url.ts:43]: @thecadams this is in `src/utils/url.ts`, which particular utility function were you referring to? Or didn't realise this was already in `utils/url`?

**Review by 0xdevalias — CHANGES_REQUESTED** [review]
(reviewing the comments I made above, I think there are at least a couple that warrant changes)

File: src/utils/url.ts (hunk 4 above)
Comment (thecadams) [reply src/utils/url.ts:43]: Pretty sure this is in `openUrl`.

Comment (denisdimitrov) [reply src/utils/url.ts:43]: This is the `openUrl` utility function, and I'm adding the check to it.

File: src/components/molecules/VenuePartygoers/VenuePartygoers.scss (hunk 5 above)
Comment (thecadams) [reply src/components/molecules/VenuePartygoers/VenuePartygoers.scss:1]: We will benefit tremendously from a documented convention for wrapper classes around components. Something like:

* A component must render a top-level element with a `className` of `COMPONENTNAME-wrapper`
* The .scss must be entirely wrapped in
```
.COMPONENTNAME-wrapper {
...
}
```

Comment (0xdevalias) [reply src/components/molecules/VenuePartygoers/VenuePartygoers.scss]: Agreed. Potentially something we can lint (`stylelint`?), if not with an existing rule, with a custom. But until then, documented/enforced convention sounds good to me.

File: src/components/molecules/NavBar/NavBar.scss [hunk 10]
```diff
@@ -282,7 +282,7 @@ $border-radius: 28px;
 
 .schedule-dropdown-body {
   position: fixed;
-  z-index: 300;
+  z-index: 4;
```
Comment (0xdevalias) [inline src/components/molecules/NavBar/NavBar.scss:285]: Use variables for these as well?

File: src/components/molecules/NavSearchBar/NavSearchBar.tsx [hunk 11]
```diff
@@ -16,7 +17,7 @@ import { NavSearchBarInput } from "./NavSearchBarInput";
 
 interface SearchResult {
   rooms: CampRoomData[];
-  users: User[];
+  users: WithId<User>[];
```
Comment (0xdevalias) [inline src/components/molecules/NavSearchBar/NavSearchBar.tsx:20]:
```suggestion
  users: readonly WithId<User>[];
```

File: src/components/molecules/Sidebar/Sidebar.scss [hunk 12]
```diff
@@ -6,7 +6,7 @@ $dark: #19181a;
 
 .sidebar-container {
   position: fixed;
-  z-index: 90;
+  z-index: 2;
```
Comment (0xdevalias) [inline src/components/molecules/Sidebar/Sidebar.scss:9]: Use variables for these as well?

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 13]
```diff
@@ -74,24 +84,23 @@ const UserProfileModal: React.FunctionComponent<PropTypes> = ({
             <div className="profile-basics">
               <div className="profile-pic">
                 <img
-                  src={fullUserProfile.pictureUrl || "/default-profile-pic.png"}
+                  src={userProfile.pictureUrl || "/default-profile-pic.png"}
```
Comment (0xdevalias) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:87]: Maybe?

```suggestion
                  src={userProfile.pictureUrl ?? "/default-profile-pic.png"}
```

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 14]
```diff
@@ -74,24 +84,23 @@ const UserProfileModal: React.FunctionComponent<PropTypes> = ({
...
+                          userProfile.id.charCodeAt(0) % RANDOM_AVATARS.length
                         )
                       ];
                   }}
                 />
               </div>
               <div className="profile-text">
                 <h2 className="italic">
-                  {fullUserProfile.partyName || "Captain Party"}
+                  {userProfile.partyName || "Captain Party"}
```
Comment (0xdevalias) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:103]: Maybe?

```suggestion
                  {userProfile.partyName ?? "Captain Party"}
```

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 15]
```diff
@@ -103,7 +112,7 @@ const UserProfileModal: React.FunctionComponent<PropTypes> = ({
                     {/*
                     // @debt typing - need to support known User interface with unknown question keys
                     // @ts-ignore */}
-                    {fullUserProfile[question.name] || //@debt typing - look at the changelog, was this a bug?
+                    {userProfile[question.name] || //@debt typing - look at the changelog, was this a bug?
```
Comment (0xdevalias) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:115]: Maybe?

```suggestion
                    {userProfile[question.name] ?? //@debt typing - look at the changelog, was this a bug?
```

Is this @debt still an issue?

**Review by 0xdevalias — COMMENTED** [review]
Comments based on changes since my last review

**Comment by denisdimitrov** [comment]
[Denis Ed tweaks](https://trello.com/c/SqX7QyaO/434-denis-ed-tweaks)

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 15 above)
Comment (denisdimitrov) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx:117]: Yes, it is.

**Comment by denisdimitrov** [comment]
@thecadams @0xdevalias You can take a second look at the changes. They've been updated.

File: src/components/organisms/UserProfileModal/UserProfileModal.scss [hunk 16]
```diff
@@ -35,8 +35,9 @@ $modal-max-width: 540px;
   }
 
   .private-chat-container {
-    height: 300px;
-    overflow-wrap: break-word;
```
Comment (thecadams) [inline src/components/organisms/UserProfileModal/UserProfileModal.scss:39]: Can we re-add this one? It makes sure we wrap by split-
ting words.

File: src/utils/url.ts (hunk 4 above)
Comment (thecadams) [reply src/utils/url.ts:43]: OK. Just thinking we should pull everything into `openUrl` including the `isExternalUrl` check. If opening URLs is in one place we can easily change the systemwide policy for whether to open in new window.

**Review by thecadams — COMMENTED** [review]
Couple new questions, looking good so far.

File: src/components/molecules/VenuePartygoers/VenuePartygoers.scss (hunk 5 above)
Comment (0xdevalias) [reply src/components/molecules/VenuePartygoers/VenuePartygoers.scss]: @thecadams RE: the above, I have some good patterns I figured out with how we can use CSS modules (enabled by default in `create-react-app` + some other improvements to our styling. Started documenting my context in https://github.com/sparkletown/internal-sparkle-issues/issues/4 the other day to avoid the rabbithole, but the main relevant part to this would be:

- https://create-react-app.dev/docs/adding-a-css-modules-stylesheet/

You can use generic names like `.error` and similar in the component's css module, then import it like `import styles from './Button.module.css'`, which then allows you to use it like `<button className={styles.error}>Error Button</button>;`

It generates locally scoped unique classnames, so no global clashes, and isn't as heavy as a lot of CSS-in-JS patterns/libs.

File: src/scss/constants.scss [hunk 17]
```diff
@@ -40,3 +40,22 @@ $playa-avatar-size: 48px;
...
+);
+
+@function z($layer) {
+  @if not map-has-key($z-layers, $layer) {
+    @warn "No layer found for `#{$layer}` in $z-layers map. Property omitted. Falling back to 1.";
+    @return 1;
+  }
+
+  @return map-get($z-layers, $layer);
+}
```
Comment (0xdevalias) [inline src/scss/constants.scss:61]: 🎉

**Review by 0xdevalias — APPROVED** [review]
LGTM! Nice addition of the `z-profile` function!

File: src/components/organisms/UserProfileModal/UserProfileModal.scss (hunk 16 above)
Comment (denisdimitrov) [reply src/components/organisms/UserProfileModal/UserProfileModal.scss:39]: I just added it back.
